src/rust/engine/async_semaphore/src/lib.rs

```diff
@@ -32,11 +32,20 @@ use std::sync::Arc;
 use futures::future::Future;
 use futures::task::{self, Task};
 use futures::{Async, Poll};
+use log::trace;
 use parking_lot::Mutex;
 
+struct UniqueTask {
+  id: usize,
+  task: Task,
+}
+
 struct Inner {
-  waiters: VecDeque<Task>,
+  waiters: VecDeque<UniqueTask>,
   available_permits: usize,
+  // Used as the source of task_id in UniqueTask's because
+  // it is monotonically increasing, and only incremented under the mutex lock.
+  task_id: usize,
```

Comment:
```suggestion
  next_task_id: usize,
```

To make it clear whether this is the most-recently-allocated or next-to-be-allocated task id :)